### ISSUE TITLE
Make layer state easily JSON-serializable

### ISSUE THREAD

**Issue by manzt** [napari/napari#1584]
## 🚀 Feature

I've been experimenting with serializing the viewer state as zarr, using `layer.as_layer_data_tuple()` to get the layer data, state, and type. (Love this method!!) However, the meta dict is not easily jsonify-able. For example `scale` and `translate` for the image layer are of type `List[np.generic]` rather than `List[float]` or `List[int]` which means you can't just dump the metadata into `attrs` easily and need to inspect each key/val to create a serializable object.

https://github.com/napari/napari/blob/2ff87e56aac2e1c812e82fa5bbf8bcf1b7954150/napari/layers/base/base.py#L497

For something simple like `scale` and `translate`, is there any reason `scale.tolist()` couldn't be used here instead? Colormaps and more complex objects are another ballgame completely, but hopefully #1575 will help in this regard. 

## Pitch

It would be ideal to have the option to get the metadata as JSON from `layer.as_layer_data_tuple()`. Maybe something like `layer.as_layer_data_tuple(as_json=True)`. This would mean making some decisions around how to express more complex layer state attributes in JSON, but then writing plugins or helper functions to create those python objects later from the JSON would be straightforward.

**Comment by sofroniewn** [comment]
> For something simple like scale and translate, is there any reason scale.tolist() couldn't be used here instead?

That would be fine, a PR would be most welcome.

> Colormaps and more complex objects are another ballgame completely, but hopefully #1575 will help in this regard.

Yeah, it will now be very easy to turn the colormaps/ color arrays into json as we have complete control over them and knowledge of what they are.

> It would be ideal to have the option to get the metadata as JSON from layer.as_layer_data_tuple(). Maybe something like layer.as_layer_data_tuple(as_json=True)

I like this idea a lot, and it could solve some of the discussions we were having about if `_get_state` should return things like the `colormap` object or things like a serializable object. I think the answer is it should be able to do both as plugins like `napari-svg` probably prefer the colormap object, but applications like serialization probably prefer a dictionary of basic types.

**Comment by manzt** [comment]
> I think the answer is it should be able to do both as plugins like `napari-svg` probably prefer the colormap object, but applications like serialization probably prefer a dictionary of basic types.

Yup, same thought. It makes sense to keep state objects as they are if interacting with state _within_ a python session, but for serialization, you want the ability to capture the minimal data (as basic types) required to recreate each object. 

Rather than `as_json=True` maybe something like `serializable=True` which means a dict of basic types. For any state object which is not composed of basic types, it might be useful to implement a common method for serializing (e.g. `state_obj.as_dict()`).

**Comment by kevinyamauchi** [comment]
I had a chat with @manzt today and it sounds like being able to serialize the layer state would really open the door to experimenting with sharing/remote napari (e.g., #1875 and as described in the top of this issue). I went ahead and flagged the state attributes that are not currently serializable below. I think it breaks down into a few reasonably-sized PRs, starting with #1940, which I am already working on. I have a good chunk of time that I was planning to dedicate to napari in December, so I'd be happy to work on the rest of them as well. What do you think, @sofroniewn ?

- [ ] Fix the color-related issues in #1940 by adding a `serializable` parameter to `ColorManager._get_state()`
- [ ] Fix the text-related issues by adding a `serializable` parameter to `TextManager._get_state()`
- [ ] Fix the scale/translate in the base layer and add a function to convert the properties to dict with lists. also, add a `serializable` parameter to all `layer._get_state()` methods

edit: I just saw that `as_dict` was just added to the `evented_dataclass` in #1966, which might be useful for `ColorManager` and `TextManager`

**Layers that have properties that prevent us from serializing the layer state:**
- [Base layer](https://github.com/napari/napari/blob/2ff87e56aac2e1c812e82fa5bbf8bcf1b7954150/napari/layers/base/base.py#L494-L501):
   - scale
   - translate

- [Labels layer](https://github.com/napari/napari/blob/8c720b350db3c1b35b4bee8b4cb572794bd7fad2/napari/layers/labels/labels.py#L387-L393)
  - `properties`. the values in the dict are numpy arrays
  - `colors`. the values in the dict are numpy arrays

- [Points layer](https://github.com/napari/napari/blob/8c720b350db3c1b35b4bee8b4cb572794bd7fad2/napari/layers/points/points.py#L1216-L1232)
    - `properties`. the values in the dict are numpy arrays
    - `*_color`. colors are numpy arrays
    - `*_color_cycle`. colors are numpy arrays
- [Shapes layer](https://github.com/napari/napari/blob/8c720b350db3c1b35b4bee8b4cb572794bd7fad2/napari/layers/shapes/shapes.py#L1240-L1256)
    - `properties`. the values in the dict are numpy arrays
    - `*_color`. colors are numpy arrays
    - `*_color_cycle`. colors are numpy arrays
- [Tracks](https://github.com/napari/napari/blob/8c720b350db3c1b35b4bee8b4cb572794bd7fad2/napari/layers/tracks/tracks.py#L215-L223)
    - `properties`. the values in the dict are numpy arrays
- [Vectors](https://github.com/napari/napari/blob/8c720b350db3c1b35b4bee8b4cb572794bd7fad2/napari/layers/vectors/vectors.py#L306-L313)
    - `properties`. the values in the dict are numpy arrays
    - `*_color`. colors are numpy arrays
    - `*_color_cycle`. colors are numpy arrays
- [Text manager](https://github.com/napari/napari/blob/8c720b350db3c1b35b4bee8b4cb572794bd7fad2/napari/layers/utils/text.py#L328-L334)
    - `color`. colors are numpy arrays
    - `translation`. this is a numpy array

**Layers that are currently serializable with caveats:**
 - [Image layer](https://github.com/napari/napari/blob/8c720b350db3c1b35b4bee8b4cb572794bd7fad2/napari/layers/image/image.py#L522-L533)
    - `colormap`. This is currently saved as the name of the colormap (string), which works for standard colormaps, but be "loadable" for custom colormaps. We should eventually add a serialization method
- [Surface layer](https://github.com/napari/napari/blob/8c720b350db3c1b35b4bee8b4cb572794bd7fad2/napari/layers/surface/surface.py#L261-L264)
    - `colormap`. This is currently saved as the name of the colormap (string), which works for standard colormaps, but be "loadable" for custom colormaps. We should eventually add a serialization method

**Comment by sofroniewn** [comment]
> I have a good chunk of time that I was planning to dedicate to napari in December, so I'd be happy to work on the rest of them as well. What do you think, @sofroniewn ?

Whoo!!!

> Layers that have properties that prevent us from serializing the layer state:

Looking at this list, is the issue anytime we have a numpy array (or even dask array?)

What is the plan for these arrays? Is that where calling `to_list` is better? What about if the array is 2D or nD? Are we going to need a special way to handle array data in this context? Or does array-like count as a "basic type" that we can leave as is. Apologies for my lack of knowledge here.

> - Fix the color-related issues in #1940 by adding a serializable parameter to ColorManager._get_state()
> - Fix the text-related issues by adding a serializable parameter to TextManager._get_state()
> - Fix the scale/translate in the base layer and add a function to convert the properties to dict with lists. also, add a serializable parameter to all layer._get_state() methods
>
> edit: I just saw that as_dict was just added to the evented_dataclass in #1966, which might be useful for ColorManager and TextManager

I think we should focus on making all our models evented_dataclasses, because we then get some serialization for free using the new `asdict` method, and get updates for free using the new `update` method. I think we'll ultimately drop the `_get_state` method. 

Looks like #1940 is already using the dataclass syntax, so I'd keep going with that there.

I'd also then pick up the `Text Manager` as evented dataclass PR that @tlambert03 started a while back, #1485. Note that no longer needs the copy of the dataclass and requirements changes, and we've renamed on or two things since then.

> colormap. This is currently saved as the name of the colormap (string), which works for standard colormaps, but be "loadable" for custom colormaps. We should eventually add a serialization method

This is now an evented_dataclass and so will have an `asdict` method.

I'm planning to pick up #1489 evented dataclass for ViewerModel after I've got a couple in progress PRs merged (will probably be a new PR).

I'm also planning to push the dims evented dataclass #1917 over the finish line soon too.

As for the individual layers themselves, it will be longer/ harder to move them over to evented dataclasses as so we might want to use the `_get_state` there a little longer.

**Comment by kevinyamauchi** [comment]
> Looking at this list, is the issue anytime we have a numpy array (or even dask array?)
>
> What is the plan for these arrays? Is that where calling to_list is better? What about if the array is 2D or nD? Are we going to need a special way to handle array data in this context? Or does array-like count as a "basic type" that we can leave as is. Apologies for my lack of knowledge here.

Most of the issues are with numpy arrays. My plan is to use the `ndarray.to_list()` method to serialize the properties that are numpy arrays. `to_list()` works for serializing to json, which is sufficient for the use cases I am aware of (e.g., [zarr attributes are json objects](https://zarr.readthedocs.io/en/stable/spec/v2.html#attributes)).

> I think we should focus on making all our models evented_dataclasses, because we then get some serialization for free using the new asdict method, and get updates for free using the new update method. I think we'll ultimately drop the _get_state method.
>
> Looks like #1940 is already using the dataclass syntax, so I'd keep going with that there.

For sure. Other than converting `TextManager` to an `evented_dataclass`I think the fixes to add a serializable mode to 


>  This is now an evented_dataclass and so will have an asdict method.

Awesome!

**Comment by kcpevey** [comment]
@kevinyamauchi were you able to make any progress on this?

**Comment by kevinyamauchi** [comment]
@kcpevey , thanks for checking in! Unfortunately, I didn't manage to make more headway on this issue.
